rfd/xxxx-bot-for-cert-renewals.md

```diff
@@ -0,0 +1,298 @@
+
+## Why
+
+Teleport nodes running the SSH service maintain a connection to the cluster and
+are able automatically renew their host certificates when a CA rotation takes
+place or the certificate is approaching its expiration date.
+
+Unfortunately, external systems that join a Teleport cluster (OpenSSH servers,
+databases, applications, etc.) are not under our control, and therefore do not
+have a built in mechanism to renew certificates under these circumstances. The
+workflow for working with these types of external systems is to generate
+certificates via `tctl auth sign`, copy them to the target, and manually
+configure the target system to use these certificates. This is a process that
+must be repeated whenever a certificate is about to expire. Additionally, this
+approach is likely to incur downtime during
+[CA rotations](https://goteleport.com/docs/architecture/authentication/#certificate-rotation)
+as the renewal has to be performed within the rotation's grace period in order
+to maintain connectivity.
+
+This experience can incentivize users to make decisions that conflict with
+security best practices, including:
+
+- Requesting certificates with a very large time-to-live (TTL)
+- Electing not to perform CA rotations
+
+Both of these decisions are simply delaying the inevitable - eventually, the
+certificates will need to be renewed and external systems will need to be
+updated. The agent proposed in this RFD intends to automate the process of
+renewing these certificates - requiring less manual work on behalf of our users,
+and enabling them to use shorter TTLs and perform frequent CA rotations.
+
+## Details
+
+### Goals
+
+The primary user this agent is designed for is the admin who wants to leverage
+Teleport for as many systems as possible, but simply does not have the time to
+keep track of all certificates and manually renew them.
+
+While users with sophisticated automation and dedicated credential management
+tools may benefit from this solution, they likely already have the resources to 
+automate certificate renewals.
+
+As a result, a primary goal for this tool is simplicity. We will optimize the user
+experience for those looking to add systems to their cluster quickly and easily rather
+than supporting a multitude of configuration options and hooks for customization.
+
+### Overview
+
+We will develop a new lightweight tool, `tbot`, which runs as an agent on
+external systems that are part of a Teleport cluster (a VM running a Postgres
+database, or an OpenSSH server, for example).
+
+This agent will join the Teleport cluster, request certificates, store them in a
+user-specified [destination](#destination), and then monitor the certificates
+and renew them as necessary.
+
+#### User Experience
+
+First, a token will be generated for the bot to join the cluster.
+
+```
+$ tctl tokens add --type=bot
+The invite token: 13b74f49d27536dd5c514073097c197b
+This token will expire in 60 minutes
```

Comment:
Note that [simplified node joining](https://github.com/gravitational/teleport/pull/7292) will make it possible to join without short-lived token, consider it in the design, what would change?

Reply:
Interesting! The biggest issue I see here is the simplified node joining is looking for the node to join the cluster within a 5 minute TTL from when the instance is launched. This seems fine if you've got a packaged AMI with everything already pre-installed and configured, but I'm not sure it makes sense when you want to join an existing non-Teleport instance like an OpenSSH server. You'd have to copy the software down, set up a system service, and then reboot the instance in order to get it to join - is that much of an improvement?

Reply:
> packaged AMI with everything already pre-installed and configured

I agree, but that's surprisingly popular use-case on AWS. I'd guess that 80% of our users follow it.

Reply:
note: with the [IAM join method](https://github.com/gravitational/teleport/blob/master/rfd/0041-aws-node-join.md#iam-method) there will be no TTL from when the instance is launched, which should make this nicer